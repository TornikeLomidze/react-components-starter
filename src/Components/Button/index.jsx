import React from 'react'
import PropTypes from 'prop-types'
import StyledButton from './styles'

const Button = props => {

  return (
    <StyledButton
      id={props.id}
      primary={props.primary}
      secondary={props.secondary}
      danger={props.danger}
      info={props.info}
      success={props.success}
      full={props.full}
      loading={props.loading}
      disabled={props.disabled || props.loading}
      className={`button-component ${props.className ? props.className : ''}`}
      xs={props.xs}
      sm={props.sm}
      md={props.md}
      lg={props.lg}
      nowrap={props.nowrap}
      type="button"
      onClick={() => {
        props.onEvent({
          type: "OnClick",
          origin: "Button",
          props: {
            data: props.data
          }
        })
      }}>
      {
        props.loading && props.loadingLabel ? props.loadingLabel : props.label
      }
    </StyledButton>
  )
}

Button.defaultProps = {
  disabled: false,
  danger: false,
  success: false,
  info: false,
  nowrap: false,
  secondary: false,
  className: '',
  loadingLabel: '',
  primary: false,
}

Button.propTypes = {
  primary: PropTypes.bool,
  secondary: PropTypes.bool,
  info: PropTypes.bool,
  danger: PropTypes.bool,
  success: PropTypes.bool,
  nowrap: PropTypes.bool,
  disabled: PropTypes.bool,
  onEvent: PropTypes.func.isRequired,
  className: PropTypes.string,
  label: PropTypes.string.isRequired,
  loadingLabel: PropTypes.string,
}

export default Button
