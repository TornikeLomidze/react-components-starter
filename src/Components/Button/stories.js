
import React from 'react';
import { storiesOf } from '@storybook/react';
import Container, { Row, Column } from '../Grid'
import Button from '.';

storiesOf('Components|Buttons', module)

  .add('Primary', () => (
    <Container>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default Primary</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button primary onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button primary loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button primary disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default Primary xs</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button primary xs onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button primary xs loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button primary xs disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default Primary sm</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button primary sm onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button primary sm loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button primary sm disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default Primary md</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button primary md onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button primary md loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button primary md disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default Primary lg</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button primary lg onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button primary lg loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button primary lg disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>



      <Row>
        <Column fullSize>
          <p>Size Full</p>
          <Button primary full onEvent={e => {
          console.log('onEvent', e)
          }} loadingLabel="Applying..."  label="Primary XS" />

          <p>Size Full Disabled</p>
          <Button primary disabled full onEvent={e => {
          console.log('onEvent', e)
          }} loadingLabel="Applying..."  label="Primary XS" />

          <p>Size Full loading</p>
          <Button primary loading full onEvent={e => {
          console.log('onEvent', e)
          }} loadingLabel="Applying..."  label="Primary XS" />
        </Column>
      </Row>

    </Container>
  ))

  .add('secondary', () => (
    <Container>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default secondary</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button secondary onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button secondary loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button secondary disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default secondary xs</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button secondary xs onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button secondary xs loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button secondary xs disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default secondary sm</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button secondary sm onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button secondary sm loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button secondary sm disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default secondary md</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button secondary md onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button secondary md loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button secondary md disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default secondary lg</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button secondary lg onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button secondary lg loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button secondary lg disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column fullSize>
          <p>Size Full</p>
          <Button secondary full onEvent={e => {
          console.log('onEvent', e)
          }} loadingLabel="Applying..."  label="secondary XS" />

          <p>Size Full Disabled</p>
          <Button secondary disabled full onEvent={e => {
          console.log('onEvent', e)
          }} loadingLabel="Applying..."  label="secondary XS" />

          <p>Size Full loading</p>
          <Button secondary loading full onEvent={e => {
          console.log('onEvent', e)
          }} loadingLabel="Applying..."  label="secondary XS" />
        </Column>
      </Row>

    </Container>
  ))

  .add('danger', () => (
    <Container>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default danger</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button danger onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button danger loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button danger disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default danger xs</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button danger xs onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button danger xs loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button danger xs disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default danger sm</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button danger sm onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button danger sm loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button danger sm disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default danger md</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button danger md onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button danger md loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button danger md disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default danger lg</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button danger lg onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button danger lg loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button danger lg disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>



      <Row>
        <Column fullSize>
          <p>Size Full</p>
          <Button danger full onEvent={e => {
          console.log('onEvent', e)
          }} loadingLabel="Applying..."  label="danger XS" />

          <p>Size Full Disabled</p>
          <Button danger disabled full onEvent={e => {
          console.log('onEvent', e)
          }} loadingLabel="Applying..."  label="danger XS" />

          <p>Size Full loading</p>
          <Button danger loading full onEvent={e => {
          console.log('onEvent', e)
          }} loadingLabel="Applying..."  label="danger XS" />
        </Column>
      </Row>

    </Container>
  ))

  .add('success', () => (
    <Container>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default success</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button success onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button success loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button success disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default success xs</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button success xs onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button success xs loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button success xs disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default success sm</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button success sm onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button success sm loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button success sm disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default success md</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button success md onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button success md loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button success md disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>
      <Row>
        <Column>
          <Row>
            <Column>
              <p>Default success lg</p>
            </Column>
          </Row>
          <Row>
            <Column>
              <Button success lg onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button success lg loading onEvent={e => {
                console.log('onEvent', e)
              }} loadingLabel="Applying..."  label="Apply" />
            </Column>
            <Column>
              <Button success lg disabled onEvent={e => {
                  console.log('onEvent', e)
                }} loadingLabel="Applying..."  label="Apply" />
            </Column>
          </Row>
        </Column>
      </Row>



      <Row>
        <Column fullSize>
          <p>Size Full</p>
          <Button success full onEvent={e => {
          console.log('onEvent', e)
          }} loadingLabel="Applying..."  label="success XS" />

          <p>Size Full Disabled</p>
          <Button success disabled full onEvent={e => {
          console.log('onEvent', e)
          }} loadingLabel="Applying..."  label="success XS" />

          <p>Size Full loading</p>
          <Button success loading full onEvent={e => {
          console.log('onEvent', e)
          }} loadingLabel="Applying..."  label="success XS" />
        </Column>
      </Row>

    </Container>
  ))



