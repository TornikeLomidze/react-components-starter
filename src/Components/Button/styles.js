import styled, { keyframes }  from "styled-components";
import Color from "../Colors";

const buttonLoadingAnimation = keyframes`
  to {
    transform: rotate(360deg);
  }
`;

const generateDisablesButtonColor = props => {
  let color = 'Blue'
  switch ( true ) {
    case props.secondary:
      color = 'Orange'
      break
    case props.info:
      color = 'Blue-Grey'
      break
    case props.danger:
      color = 'Red'
      break
    case props.success:
      color = 'Green'
      break
    default:
      color = 'Blue'
  }


  return (`
    background-color: ${Color[`${color}-50`]};
    border: 1px solid ${Color[`${color}-50`]};
    color: ${props.loading ? Color[`${color}-300`] : Color[`${color}-100`]};
    cursor: ${props.loading ? 'wait' : 'not-allowed'};
    :hover {
      background-color: ${Color[`${color}-50`]};
      border: 1px solid ${Color[`${color}-50`]};
    }
`)
}

const generateButtonColor = color => `
  background-color: ${Color[`${color}-A700`]};
  border: 1px solid ${Color[`${color}-800`]};

  :hover {
    background-color: ${Color[`${color}-300`]};
    border: 1px solid ${Color[`${color}-300`]};
  }

  :active {
    background-color: ${Color[`${color}-300`]};
    border: 1px solid ${Color[`${color}-300`]};
    color: ${Color[`${color}-600`]};
  }
`

export default styled.button`
  position: relative;
  width: fit-content;
  color: ${Color["Blue-Grey-900"]};

  padding: 12px;
  border-radius: 3px;
  font-family: Roboto;
  font-size: 14px;
  line-height: auto;
  text-align: center;
  cursor: pointer;

  ${({ nowrap }) => nowrap && `
    white-space:nowrap;
  `}

  ${({ xs }) => xs && `
    font-size: 10px;
    padding: 4px;
  `}

  ${({ sm }) => sm && `
    font-size: 12px;
    padding: 6px;
  `}

  ${({ md }) => md && `
    padding: 8px;
  `}

  ${({ lg }) => lg && `
    padding: 10px;
  `}

  ${({ full }) => full && `
    width: 100%;
  `}

  ${({ primary }) => primary && `
    ${generateButtonColor('Blue')}
  `}

  ${({ secondary }) => secondary && `
    ${generateButtonColor('Orange')}
  `}

  ${({ danger }) => danger && `
    ${generateButtonColor('Red')}
  `}

  ${({ success }) => success && `
    ${generateButtonColor('Green')}
  `}

  ${props => props.disabled && `
    ${generateDisablesButtonColor(props)}
  `}



  :before {
    content: '';
    box-sizing: border-box;
    position: absolute;
    top: 50%;
    left: 50%;
    width: 0px;
    height: 0px;
    margin-top: -15px;
    margin-left: -15px;
    border-radius: 50%;
    border: 1px solid #ccc;
    border-top-color: #07d;
    animation: ${buttonLoadingAnimation} .6s linear infinite;

    ${({ loading }) => loading && `
      width: 30px;
      height: 30px;
    `}
  }

`;
