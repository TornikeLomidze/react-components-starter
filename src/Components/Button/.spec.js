import React from 'react'
import renderer from 'react-test-renderer'

import Button from '.'

describe('Button', () => {
    it('Rendering Button without crashing', () => {
        const tree = renderer.create(
          <Button
            onEvent={e => {
            }}
            label="Test Button"
          />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
})