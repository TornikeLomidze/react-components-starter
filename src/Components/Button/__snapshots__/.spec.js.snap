// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Rendering Button without crashing 1`] = `
<button
  className=""
>
  Test Button
</button>
`;
